### ISSUE TITLE
Make timer callbacks respect reentrancy of grains

### ISSUE THREAD

**Issue by sergeybykov** [dotnet/orleans#2574]
Today, timer callbacks interleave with grain methods (at `await` points) for both reentrant and non-reentrant grains. This is a source of confusion because developers, naturally, do not expect interleaving in non-reentrant grains. We need to make timer callbacks respect reentrancy of grain classes and not interleave for non-reentrant grains.

**Comment by cata** [comment]
Heads-up: this is a breaking change to an actively used feature(/bug? :-) )

We are actively relying on timer callback reentrancy. This being said, I agree that the behavior is confusing and should be changed. I think a straightforward migration path would be to call a reentrant grain method (specified using `AlwaysInterleave` or `MayInterleave`) from the timer callback.

**Comment by sergeybykov** [comment]
@cata, thanks for the heads-up. I didn't realize people rely on this behavior.

**Comment by kogir** [comment]
This timer reentrancy initially surprised me as well, but now our code has also come to rely on it. The non-reentrant methods in our grains often control flags used by the timer callback.

Would be happy to migrate, but this "loophole" is currently the only way to have a partially reentrant grain - a useful feature that would be sad to lose.

**Comment by sergeybykov** [comment]
Maybe we should make the behavior configurable then, to explicitly allow for both reentrant and non-reentrant execution of timer callbacks?

**Comment by cata** [comment]
@sergeybykov - configurable behavior (an extra parameter on a `RegisterTimer` overload?) would work great for us. 
The *default* behavior could be changed to match the grain's default reentrancy policy, thus making it less surprising.

**Comment by gabikliot** [comment]
I think that is exactly what we agreed upon and planned to do a long time ago. That is at least what I remember.

**Comment by jamescarter-le** [comment]
@sergeybykov @gabikliot Please can I clarify if this behavior is applicable to Reminders too?

**Comment by sergeybykov** [comment]
@jamescarter-le Reminders are different in behavior from timers because their ticks are delivered via true grain calls that follow reentrancy/interleaving rules.

**Comment by shlomiw** [comment]
It's very important for timers callback not to interleave. Optional configuration (to allow re-entrant) is also welcomed.
When do you plan releasing it? 

Please consider also an option to register a timer as such it wont allow Orleans deactivating the Grain.
Thanks.

**Comment by sergeybykov** [comment]
We don't have concrete plans for this. But we are open to contributions.

**Comment by shlomiw** [comment]
In the meanwhile, what is the correct way to 'self invoke' from a timer method to avoid interleaving?
using `this.AsReference<IFoo>()` ?
is it possible to achieve this without exposing the invoked method in the interface?

**Comment by gabikliot** [comment]
Yes, as reference.
No, it has to be exposed via interface. But you can define a different interface, it can be internal in the grain impl. assembly.

**Comment by shlomiw** [comment]
Ok, got it, thanks

**Comment by shlomiw** [comment]
To avoid the reentrancy of timers - I've used in my timer `this.AsReference<IFoo>()` to self-invoke a different method and by that it respects the non-reentrancy definition of the grain.
Now, I've seen that sometimes, when there are other methods which takes long time, then this method is executed multiple times sequentially and not in a 'timed' manner.

The timer only invokes the other method which get its turn scheduled, and sometimes, due to other long methods turns, it can be scheduled sequentially, and by that it doesn't respect the expected 'timer' intervals (i.e. run every x seconds).

Now, I can add a defense to avoid it running sequentially by checking the last time it was executed, but I was wondering if you have a better solution.

FYI - this is the method I used to self-invoke a method with the Grain context scheduler:
```
public static class GrainExtensions
    {
        /// <summary>
        /// SelfInvokeAfter
        /// invoke self method in next 'turn'
        /// this is how it's done to avoid interleaving and re-entrant issues
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="grain"></param>
        /// <param name="action"></param>
        public static void SelfInvokeAfter<T>(this Grain grain, Func<T, Task> action)
            where T : IGrain
        {
            var g = grain.AsReference<T>();
            var factory = new TaskFactory(TaskScheduler.Current);
            factory.StartNew(async () => await action(g));
        }
    }
```

**Comment by jason-bragg** [comment]
@shlomiw I may be misunderstanding what you're trying to do here, but if the goal is to ensure that work triggered by the timer conforms to grain call semantics (ie, not reentrant) you need only make await the grain call.

        public static Task SelfInvokeAfter<T>(this Grain grain, Func<T, Task> action)
            where T : IGrain
        {
            var g = grain.AsReference<T>();
            return action(g));
        }

Assuming "action" contains a call to the grain, awaiting SelfInvokeAfter in the timer callback will prevent the next timer call from being scheduled until the first one is complete.  By calling factory.StartNew(async () => await action(g)) and not waiting the completion of the operation, the timer callback is free to return immediately and begin counting down to the next trigger, before current work is complete.

**Comment by shlomiw** [comment]
If I can await the self-invoke from the timer then it sure makes sense (I thought I have to schedule it for next turn, to avoid 'deadlock', but it should be possible since the timer is reentrant!).
Thanks! you perfectly understood me. I'll validate it and let you know.

**Comment by shlomiw** [comment]
@jason-bragg - thanks! it works correctly as you suggested :)

**Comment by jason-bragg** [comment]
@shlomiw, Glad that helped.
That's a common pattern used by those that want to schedule periotic work that respects Orleans call semantics.
It should be noted that since the timer is generating a grain call, assuming the timer period is shorter than the grain collection time, grains using this pattern will never be deactivated due to idleness, because the timer is always generating calls, which is interpreted as activity.
I don't know your service so I won't assume this is an issue for you, but wanted you to be aware of this side effect.

**Comment by shlomiw** [comment]
@jason-bragg - I'm aware of it, thanks for bringing that up!
I still think you should let the timers be configurable about their re-entrancy, I would also say that the default should be same as the Grain. This can be very confusing and cause unexpected results.

**Comment by shlomiw** [comment]
@jason-bragg - 

> It should be noted that since the timer is generating a grain call, assuming the timer period is shorter than the grain collection time, grains using this pattern will never be deactivated due to idleness, because the timer is always generating calls, which is interpreted as activity.

Now I have a scenario when I want to use timer for grain local maintenance, which still respects the re-entrancy (i.e. - non-entrant), but I don't want it to keep the grain alive. 

Any idea how?

**Comment by sebastianburckhardt** [comment]
I am sometimes using await loops with delays:

```csharp
while (! done)
{
     do_maintenance_work();
     await Task.Delay(period);
}
```

As far as I know, this does the right thing if the grain deactivates (i.e. it will no longer resume at the await) but I am not 100% sure.

**Comment by shlomiw** [comment]
@sebastianburckhardt - tnx for your reply!
I'm not sure I understand. In what you suggested, if my grain is non-reentrant (as in my case) it would never get outside requests.

**Comment by shlomiw** [comment]
btw - instead of using timers, I was thinking about using grain call filter (interceptor) and execute my maintenance after I got a request (i.e. every once in a while).

**Comment by skyflyer** [comment]
@gabikliot, @sergeybykov, @jason-bragg,

I just tested timer invocation with the suggested workaround and I still see interleaving calls. I'm using Orleans `2.0*`.

<details>
<summary>The grain implementation</summary>

```c#
using System;
using System.Threading.Tasks;
using OrleansGrainInterfaces;
using Orleans;
using Microsoft.Extensions.Logging;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using Orleans.Concurrency;

namespace OrleansGrains
{
    public class ValueGrain : Grain, IValueGrain
    {
        private readonly ILogger _logger;
        private readonly Random _rnd;
        IDisposable _timer;

        public ValueGrain(ILogger<ValueGrain> logger) {
            _logger = logger;
            _rnd = new Random();
        }

        public async Task<string> Get()
        {
            Console.WriteLine($"***Start***");
            await Task.Delay(3000);
            
            Console.WriteLine($"***Finish***");
            return this.GetPrimaryKeyString();
        }

        public async Task OnTimer(object data)
        {
            await this.AsReference<IValueGrain>().Update();
        }

        public async Task Update()
        {
            Console.WriteLine($"***StartTimer***");
            var waitPeriod = _rnd.Next(5000,15000);
            await Task.Delay(waitPeriod);
            Console.WriteLine($"***FinishTimer***");
        }

        public override Task OnActivateAsync()
        {
            _timer = RegisterTimer(this.OnTimer, null, TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(10));
            return Task.CompletedTask;
        }

        public override Task OnDeactivateAsync()
        {
            _timer?.Dispose();
            return Task.CompletedTask;
        }
    }
}
```
</details>

Output produced:

```
***Start***
***StartTimer*** <-- should not start, since the `Get` method was not finished yet
***Finish***
***Start*** <-- new `Get` invocations during a timer invocation
***Finish***
***Start***
***FinishTimer***
***Finish***
```

I understand that the cals are (supposedly?) interleaved between `await`, but the [documentation](https://dotnet.github.io/orleans/Documentation/grains/timers_and_reminders.html) explicitly states (emphasis mine):

> Each invocation of asyncCallback ***is delivered to an activation on a separate turn and will never run concurrently with other turns on the same activation***. Note however, asyncCallback invocations ***are not delivered as messages*** and are thus not subject to message interleaving semantics. This means that invocations of asyncCallback should be considered to behave as if running on a reentrant grain with respect to other messages to that grain.

I guess I'm not understanding this two sentences correctly, as they seem contradictory to me.

So, timer callbacks will always interleave (based on testing), but the workaround, by using `this.AsReference<IValueGrain>().Update()` should honor the turn-based method invocation, but it does not in practice.

I'd appreciate an explanation, as I'm probably just misunderstanding some concepts here.

**Comment by galvesribeiro** [comment]
@sergeybykov 

I know people is now relying on this _bug_ as previous comments but, it is indeed not the correct behaviour. 

We are having issues now with this behaviour... 

Will this be ever fixed?

**Comment by brianqcgames** [comment]
@skyflyer 

I am having this same issue after upgrading to Orleans 2.0 and the call chain reentrancy is the issue. Call chain reentrancy will interleave for self calls, A->B->A calls, and stream messages.

Despite the documentation, timers have never respected the grain scheduler and with 2.0 we no longer have a viable workaround.

My only option (unless someone has a better idea) is to disable the call-chain reentrancy so my original workaround functions again. Fortunately, since we wrote that vast majority of our code prior to 2.0 we already accounted for the deadlock issues present without call chain reentrancy.

**Comment by shlomiw** [comment]
@skyflyer - wow, actually scarry stuff, this could lead to unpredictable issues in my current implementation. Thanks for the update! 
I'm still using orleans 2.0-beta3 (wanted to upgrade soon). I'll test it as well.

**Comment by shlomiw** [comment]
@skyflyer 
One **very ugly hack** you can try - use the new `HostedClient` introduced in 2.1 to call your grain interface from the timer method. It should respect the reentrancy. I don't know about the performance hit implications and if it's relevant for you.

**Comment by galvesribeiro** [comment]
@shlomiw you mean inject IClusterClient (after enable the hosted client implementation with `.EnableDirectClient()`) on a grain and invoke it?

If that is an yes, than yeah... That is a **_VERY_** ugly hack heheheh

**Comment by ifle** [comment]
We have a lot Orleans.Storage.InconsistentStateException errors in non reentrancy grains with timer under heavy load.
Looks like the timer interleaving issue
```c#
Error from storage provider AdoNetGrainStorage.Server.Grains.ProjectGrain 
during WriteState for grain 
Type=Server.Grains.ProjectGrain Pk=*grn/9955FAF8/0000000000000000000000000000001703ffffff9955faf8-0xFBF1FF54 
Id=GrainReference:*grn/9955FAF8/00000017 Error=
 
 Exc level 0: Orleans.Storage.InconsistentStateException: Version conflict (WriteState): 
 ServiceId=pickpack-app ProviderName=Default GrainType=Server.Grains.ProjectGrain GrainId=23 ETag=185.
 at Orleans.Storage.AdoNetGrainStorage.<WriteStateAsync>d__33.MoveNext()
 --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at Orleans.Core.StateStorageBridge`1.<WriteStateAsync>d__12.MoveNext()	

```

**Comment by ifle** [comment]
@sergeybykov  Please change this interleaving bug to feature (fix documentation ) and add the flag that will make timer callbacks respect reentrancy of grains.

**Comment by shlomiw** [comment]
FYI - with orleans 2.0-beta3 the `AsReference()` trick does respect the re-entrancy. When we'll find time to upgrade to 2.2, I'll test it again.

Until then - I believe it would be better for all that timer will just respect the re-entrancy of the grains as default. I think it's far more dangerous otherwise..

**Comment by skyflyer** [comment]
Perhaps this is specific to 2.0-beta3 or perhaps I'm not using it right, but I tried `AsReference` in callback handler as well as in timer registration and my timer calls still interleave with client calls. I was testing this with 2.1.2.

**Comment by jonas-johansson** [comment]
Our team is in a tough spot. Our code has come to rely on call chain reentrancy but we'd like to execute code periodically (time span of seconds). Timers seems like a good fit except for the disrespect for reentrancy. We too would like to have the behavior stated in the original post:
> We need to make timer callbacks respect reentrancy of grain classes and not interleave for non-reentrant grains.

**Comment by jonas-johansson** [comment]
Is there some trick to execute code from an async timer callback and have it respect the reentrancy?

**Comment by Cephei** [comment]
I believe I'm also running into this issue but my knowledge of Orleans is quite limited. I have a non-reentrant grain with a few `[AlwaysInterleave]` methods. The method in question where the bulk of the processing is done is **not** marked with the aforementioned attribute. Within in the `ProcessRequestAsync()` I am executing this:

```csharp
... 
// below is not awaited
Task.Factory.StartNew(
    async () => 
    {
        // while no cancellation has been requested AND the strategy
        // has not completed, periodically save the state.
        // IsCompleted here does not mean success, but may mean cancelled as well
        while (!_cancellationTokenSource.IsCancellationRequested && 
               !strategy.Completion.IsCompleted)
        {
            await PersistState(strategy.GetSnapshot()); // this is where I receive the exception
            await Task.Delay(strategy.PersistStateInterval);
        }
    }, 
    CancellationToken.None,
    TaskCreationOptions.None,
    TaskScheduler.Current);

return strategy.Completion
    .ContinueWith(_ => PersistState(strategy.GetSnapshot()), TaskScheduler.Current)  // allows the task to save the final grain state using the Orleans TaskScheduler
    .Unwrap();
```
```csharp
protected internal virtual Task PersistState(StateSnapshot stateSnapshot)
{
    State.Data = stateSnapshot;
    return WriteStateAsync();
}
```

**Comment by Rohansi** [comment]
Has there been any progress with this? I'm narrowing down a deadlock and the only thing in common is my timer code gets blocked with something else.

@jonas-johansson Is your fork working? What are the effects of your changes?

**Comment by jonas-johansson** [comment]
@Rohanski: We're using my fork in production and it works for our case. The fork allows the call chain reentrance behavior to be configured so that self-reentrancy is not allowed (A-A), while still allowing call chain reentrancy when calling another grain (A-B-A). This has the desired effect that we in a timer callback can execute this.AsReference<IMyGrain>(...).DoSomething () and DoSomething will respect interleaving.

**Comment by Rohansi** [comment]
`Task.Run` seems to reset the call chain so to workaround this with `AllowCallChainReentrancy = true`  we just need to use the `AsReference` hack in `Task.Run`.

```csharp
private async Task TimerCallback(object arg)
{
    await Task.Run(() => this.AsReference<TGrainInterface>().ActualTimerCallback());
}
```

The actual call will occur on the proper Orleans worker thread still, not the thread pool.

**Comment by Xanatus** [comment]
@Rohansi if this works you made my day. After months wrestling with getting timers to behave correctly, this would be finaly a good enough solution for me.

I'm just not sure how it works and if its reliable enough, considering Task.Run is leaving the orleans context and scheduler entirely and I though grain calls wouldn't be possible in this case.

**Comment by jonas-johansson** [comment]
@Rohansi and @Xanatus: When I invoke the method like you did (from the action in Task.Run) I get the following exception:
```
System.InvalidOperationException: 'Support for accessing the Orleans runtime from a non-Orleans context is not enabled. Enable support using the ISiloHostBuilder.EnableDirectClient() method.'
```
Have you enabled DirectClient @Rohansi?

**Comment by Rohansi** [comment]
@jonas-johansson Yeah, I have it enabled. I think OrleansDashboard enables it if anyone is using it.

**Comment by shlomiw** [comment]
@jonas-johansson 
> System.InvalidOperationException: 'Support for accessing the Orleans runtime from a non-Orleans context is not enabled. Enable support using the ISiloHostBuilder.EnableDirectClient() method.'

You need to invoke the grain reference via Orleans Task Scheduler. `Task.Run` will invoke it in the .Net ThreadPool Task Scheduler context.

i.e. something like that:

```
public static class GrainExtensions
    {
        /// <summary>
        /// SelfInvokeAfter
        /// invoke self method in next 'turn'
        /// this is how it's done to avoid interleaving and re-entrant issues
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="grain"></param>
        /// <param name="action"></param>
        public static void SelfInvokeAfter<T>(this Grain grain, Func<T, Task> action)
            where T : IGrain
        {
            var g = grain.AsReference<T>();
            var factory = new TaskFactory(TaskScheduler.Current);
            factory.StartNew(async () => await action(g));
        }
    }
```

**Comment by Xanatus** [comment]
@shlomiw your approach isn't working with `AllowCallChainReentrancy`, which is what the discussion was about.

**Comment by shlomiw** [comment]
@Xanatus - oh yes, sorry, now I see.

**Comment by shlomiw** [comment]
@sergeybykov - can you reconsider this issue again? let timers respect the reentrancy? (make it configurable)..
As you can see it's a source of a lot of confusion, unexpected behaviour and weird workarounds.

**Comment by sergeybykov** [comment]
@shlomiw By "reconsider" do you mean "raise the priority" of this issue?

**Comment by Xanatus** [comment]
@sergeybykov  I use orleans for over 2 years now and for me timers always have been the number 1 source of problems, making development extremely difficult and hacky for something very basic that really shouldn't be so complicated. Especially now that we don't have a proper workaround anymore, I believe this issue would deserves some top priority.
This thread also makes clear that the issue is neither new, nor rare. It just feels like the issue is beeing ignored for the wrong reasons. ;)

**Comment by shlomiw** [comment]
> @shlomiw By "reconsider" do you mean "raise the priority" of this issue?

@sergeybykov - oh yes please 👍

**Comment by sergeybykov** [comment]
Can't promise anything in the near term as we have quite a bit of work scheduled already. Contributions, of course, are very welcome.

**Comment by ErroneousFatality** [comment]
Just reporting in to say that this issue caused me 2 work days of debugging and testing before I found out this thread, and then found out there's no sensible fix without being hacky, to my sorrow.
Hope this gets fixed soon.

**Comment by jonas-johansson** [comment]
I just wanted to give an update: our team is now running with DirectClient enabled and we're using the trick from @Rohansi:

> 
> 
> `Task.Run` seems to reset the call chain so to workaround this with `AllowCallChainReentrancy = true` we just need to use the `AsReference` hack in `Task.Run`.
> 
> ```cs
> private async Task TimerCallback(object arg)
> {
>     await Task.Run(() => this.AsReference<TGrainInterface>().ActualTimerCallback());
> }
> ```
> 
> The actual call will occur on the proper Orleans worker thread still, not the thread pool.

It works great for us!

**Comment by Galileon** [comment]
yo is there any update about this issue ? I still cannot make it working properly with provided solutions :(

**Comment by Galileon** [comment]
> I just wanted to give an update: our team is now running with DirectClient enabled and we're using the trick from @Rohansi:
> 
> > `Task.Run` seems to reset the call chain so to workaround this with `AllowCallChainReentrancy = true` we just need to use the `AsReference` hack in `Task.Run`.
> > ```cs
> > private async Task TimerCallback(object arg)
> > {
> >     await Task.Run(() => this.AsReference<TGrainInterface>().ActualTimerCallback());
> > }
> > ```
> > 
> > 
> > The actual call will occur on the proper Orleans worker thread still, not the thread pool.
> 
> It works great for us!

yea it works but is it really safe from perfomance perspective :) ?

**Comment by JorgeCandeias** [comment]
Just leaving a note that some patterns, e.g. reactive caching / reactive polling, *depend* on timers being re-entrant, so others calls are not blocked by long-running background tasks. While I generally agree that timers should be non-reentrant *by default*, if this now becomes true, then please allow some opt-in mechanism for the current always-reentrant behaviour. The current behaviour is very useful in the appropriate context, and allows for very sleek asynchronous algorithms.

**Comment by OracPrime** [comment]
> 
> 
> I just wanted to give an update: our team is now running with DirectClient enabled and we're using the trick from @Rohansi:
> 
> > `Task.Run` seems to reset the call chain so to workaround this with `AllowCallChainReentrancy = true` we just need to use the `AsReference` hack in `Task.Run`.
> > ```cs
> > private async Task TimerCallback(object arg)
> > {
> >     await Task.Run(() => this.AsReference<TGrainInterface>().ActualTimerCallback());
> > }
> > ```
> > 
> > 
> > The actual call will occur on the proper Orleans worker thread still, not the thread pool.
> 
> It works great for us!

I don't understand why, but I tried this approach and was getting unexpected interleaving.  Whereas if I use the AsReference but not the Task.Run then I get the desired non-interleaved result.  Which is not what I expect after reading this issue, but at least it's a good result.

**Comment by Ilchert** [comment]
We have some tests for such behavior and in version 3.4 it works as expected. @ReubenBond, could you confirm that it was fixed?

**Comment by zahirtezcan-bugs** [comment]
With 3.4 we had similar issue with timer callbacks which implies interleaving

```csharp
// Grain function
public async Task Done()
{
    if (resource == null) return;

    // nullify via Task.Run
    await Task.Run(() =>{
       resource.Done();
       resource = null;
    });

    logger.LogInformation("We are done");
}

//Timer callback
public async Task Callback()
{
    if (resource == null) return;

    // resource should not be null here
    resource.Verify(); //<-- null reference exception

    // no-await until here
    await Notify();
}

```

**Comment by E-Rockalanche** [comment]
Any progress on this? Just found out about this issue today. My team thought timers respected grain reentrancy too. Unfortunately I can't get the suggested workaround to work in 3.4.2

**Comment by Rohansi** [comment]
> Any progress on this? Just found out about this issue today. My team thought timers respected grain reentrancy too. Unfortunately I can't get the suggested workaround to work in 3.4.2

I'm on 3.4.3 and the workaround still looks like it's working for me. Here's all the pieces:

```csharp
// in your grain implementation
_updateTimer = RegisterTimer(UpdateCallback, null, UpdateInterval, UpdateInterval);
```

```csharp
// in your grain implementation
private async Task UpdateCallback(object arg)
{
    await Task.Run(() => this.AsReference<IDynamicEntity>().UpdateInternal());
}
```

```csharp
// in your grain interface
public interface IDynamicEntity : IGrainWithGuidKey
{
    // other stuff from your grain interface...

    /// <summary>
    /// For internal use only. The callback for the update timer.
    /// </summary>
    ValueTask UpdateInternal();
}
```

```csharp
// in your grain class
public async ValueTask UpdateInternal()
{
    // do the stuff you need your timer to do
}
```

**Comment by E-Rockalanche** [comment]
> await Task.Run(() => this.AsReference<IDynamicEntity>().UpdateInternal());

Oh I see. Invoking the grain in the RegisterTimer callback still ignores reentrancy. If found using regular timers works too. Is there any drawback to doing it this way?

`_updateTimer = new Timer( state => this.AsReference<IDynamicEntity>().UpdateInternal(), null, UpdateInterval, UpdateInterval);`

**Comment by Rohansi** [comment]
> I found using regular timers works too. Is there any drawback to doing it this way?
> 
> `_updateTimer = new Timer(arg => this.AsReference<IDynamicEntity>().UpdateInternal(), null, UpdateInterval, UpdateInterval);`

Calls could overlap when not using the Orleans timers. See more here: https://dotnet.github.io/orleans/docs/grains/timers_and_reminders.html

Also one thing to keep in mind with this workaround is unlike stock Orleans timers this will keep your grain active because it's issuing a grain call. Plain Orleans timers will not keep your grain active.

**Comment by ReubenBond** [comment]
I think we should address this. There are several options to consider:

1. Do we create a new timer type which respects reentrancy or add a parameter to the existing `RegisterTimer` method? A parameter is more user-friendly, in my opinion.
1. Do we change the default behavior? I think **yes**, since this is what I believe that people expect.
1. Do we continue to allow always-reentrant timers (the current default)? I think yes

Input more than welcome. The workaround is fine (it's robust and clear), but adds boilerplate and as @Xanatus indicated by commenting on #7099, it's a clear papercut.

**Comment by kwaazaar** [comment]
> You need to revert to the old 1.x behavior by disabling call chain reentrancy
> `.Configure<SchedulingOptions>(options => { options.AllowCallChainReentrancy = false; })`
> After you do that it will work as expected again using the `AsReference` trick.

When I do this, combined with the timer implementation `await this.AsReference<IMyGrain>().TimerStuff()`, my reminders no longer work. I need the reminder to reactivate my grain after getting deactivated (and then reregister the timer again).

**Comment by Rohansi** [comment]
> > You need to revert to the old 1.x behavior by disabling call chain reentrancy
> > `.Configure<SchedulingOptions>(options => { options.AllowCallChainReentrancy = false; })`
> > After you do that it will work as expected again using the `AsReference` trick.
> 
> When I do this, combined with the timer implementation `await this.AsReference<IMyGrain>().TimerStuff()`, my reminders no longer work. I need the reminder to reactivate my grain after getting deactivated (and then reregister the timer again).

Not sure why that would break your reminders but this is not necessary when using the workaround.

> Input more than welcome.

IMO the timer default should be consistent with the grains as a whole - no reentrancy unless opted in. Allowing reentrancy can make grain code so much more complicated than it needs to be. A parameter on `RegisterTimer` to opt-in to reentrancy would be perfect.

**Comment by cata** [comment]
> 1. Do we create a new timer type which respects reentrancy or add a parameter to the existing `RegisterTimer` method? A parameter is more user-friendly, in my opinion.
> 2. Do we change the default behavior? I think **yes**, since this is what I believe that people expect.
> 3. Do we continue to allow always-reentrant timers (the current default)? I think yes
> 
> Input more than welcome. [...]

I think the proposed answers (adding a parameter to `RegisterTimer`, timer reentrancy default in line with user expectations, and keeping the option to have reentrant timers) are sound (and in line with the initially envisioned approach - see the conversation preceding [this comment](https://github.com/dotnet/orleans/issues/2574#issuecomment-271787894)).
 
I would like to restate that answering *yes* to point 3 is very important to us and others (as per comment history) that have come to rely on the always-reentrant timer behavior.

**Comment by JorgeCandeias** [comment]
Note that changing the default behaviour of timers, without changing the signature of `RegisterTimer` will be an invisible breaking change and so a dangerous one for existing code upgrading package versions. I've done plenty of client work using some advanced patterns that *depend* on interleaving timers. If the default behaviour is to change, I suggest adding some kind of way to notify developers upon build. Either by making the opt-in flag parameter in the `RegisterTimer` method required, so old code breaks outright, or by adding an analyzer warning to suggest developers use the `RegisterTimer` signature with the flag. The developers can always suppress it, if not important for them.

**Comment by remcoros** [comment]
This just hits us hard. Spend the whole day figuring out why we had unexpected interleaving calls. Thinking we did something wrong (like forgetting an await or doing weird things with Task.Run).

Only to find this issue and read that timers interleave by default???

This needs some serious note in the documentation.

**Comment by Galileon** [comment]
hi finally only this one works:

await Task.Factory.StartNew(() => this.AsReference<ITestGrain>().TestGrainState($"(!!!Timer!!!) {DateTime.Now.Millisecond}")).Unwrap(); // Tak działa

Cheers
Rafal

**Comment by feitzi** [comment]
Any update on this topic? It seems that the Timer still ignores the reentrance philosophy of grains.

**Comment by ReubenBond** [comment]
We plan to address this in/before Orleans 8.0

**Comment by turowicz** [comment]
So what is the prescribed solution to have a thread safe invocation?

**Comment by Galileon** [comment]
hi, cześć
This works

async Task OnCheckDirectedInteractionsTimeOut(object state) => await Task.Factory.StartNew(() => this.AsReference<IInteractionsRouterGrain>().ProcessOnCheckDirectedInteractionsTimeOut(), CancellationToken.None, TaskCreationOptions.PreferFairness, _ts);

      public InteractionsRouterGrain(ILogger<InteractionsRouterGrain> logger,
            Extent.Services.Base.Perfomance.ServiceDiagnosticsCounters diagnosticCounters,
            SessionServiceConfiguration sessionServiceConfiguration,
            Extent.Services.MediaService.Proto.MediaService.MediaServiceClient mediaServiceClient,
            IOrleansContextAccessor contextAccessor)
        {
            (_log, _sessionServiceConfiguration, _contextAccessor, _diagnosticCounters) = (logger, sessionServiceConfiguration, contextAccessor, diagnosticCounters);
            _ts = TaskScheduler.Current;
        } 

 public async override Task OnActivateAsync()
        {
            _log.LogDebug("Activating for {InstanceId} / {GrainIdentityString}...", this.GetPrimaryKey(), IdentityString);
            var streamProvider = GetStreamProvider(OrleansConstants.SessionServiceStreamProviderName);
            await streamProvider.GetStream<UserSessionStateChangedGrainEvent>(this.GetPrimaryKey(), OrleansConstants.UserSessionStateChangedStreamNamespace).SubscribeAsync(OnNextAsync);
            await streamProvider.GetStream<InteractionQueuedGrainEvent>(this.GetPrimaryKey(), OrleansConstants.InteractionsQueuedStreamNamespace).SubscribeAsync(OnNextAsync);
            await streamProvider.GetStream<ServiceSettingChangedGrainEvent>(this.GetPrimaryKey(), OrleansConstants.ServiceSettingChangedStreamNamespace).SubscribeAsync(OnNextAsync);
            await streamProvider.GetStream<InteractionMediaCallDisconnectedGrainEvent>(this.GetPrimaryKey(), OrleansConstants.InteractionMediaCallDisconnectedStreamNamespace).SubscribeAsync(OnNextAsync);
            await streamProvider.GetStream<InteractionMediaCallConnectedGrainEvent>(this.GetPrimaryKey(), OrleansConstants.InteractionMediaCallConnectedStreamNamespace).SubscribeAsync(OnNextAsync);
            await LoadServiceSettings();
            RegisterTimer(OnCheckDirectedInteractionsTimeOut, null, TimeSpan.Zero, TimeSpan.FromSeconds(5));
            RegisterTimer(OnReditectInteractionTimerMethod, null, TimeSpan.Zero, TimeSpan.FromSeconds(5));
            await base.OnActivateAsync();
        }


i didnt found anything better since that time .

**Comment by turowicz** [comment]
This seems to work:

```
        public override async Task OnActivateAsync(CancellationToken cancellationToken)
        {
            _ = RegisterTimer(FlushTimerAsync, null, timeout, timeout);

            await base.OnActivateAsync(cancellationToken);
        }

        public async Task FlushAsync()
        {
            // activation safe code
        }

        private async Task FlushTimerAsync(object arg)
        {
            await Task.Run(async () => await this.AsReference<IMyGrain>().FlushAsync());
        }
```

**Comment by ReubenBond** [comment]
@turowicz You should be able to remove the `Task.Run` part of this and await the grain call instead.
```csharp
private async Task FlushTimerAsync(object arg)
{
    await this.AsReference<IMyGrain>().FlushAsync();
}
```

**Comment by turowicz** [comment]
@ReubenBond in my tests I get `System.InvalidOperationException : Activation access violation. A non-activation thread attempted to access activation services.` with that

**Comment by Rohansi** [comment]
Would really love to see this be resolved so grain timers fit within the actor's expected concurrency model.

**Comment by turowicz** [comment]
@ReubenBond this is the failing change: 

Commit: https://github.com/Surveily/Orleans.Streaming.Grains/commit/1cdd9e5c7203b056b21f8702b12d6707eff87a47
Build: https://github.com/Surveily/Orleans.Streaming.Grains/actions/runs/5782155410

**Comment by turowicz** [comment]
Actually the build fails with a different test now. This must be PhaseOfTheMoon. I will fix the error and see if the access violation shows.

**Comment by turowicz** [comment]
OK so it fails correctly, the access violation is swallwed inside the IQueueAdapterReceiver, but it shows in test as invalid number of invocations in Moq due to reattempt to process message.

**Comment by turowicz** [comment]
@ReubenBond it fails correctly as the access violation is swallowed inside the IQueueAdapterReceiver, but it shows in test as invalid number of invocations in Moq due to reattempt to process message.

**Comment by yevhen** [comment]
When I want timer to execute in non-interleaved way I just call self via interface method. Works like a charm.

**Comment by turowicz** [comment]
@yevhen thats great but my calls are made via IQueueAdapter and it throws Access Violations.

**Comment by yevhen** [comment]
It shouldn't make any difference in theory. AFAIU, the internal client just skips the socket and routes messages by directly using the silo's internal infrastructure.

**Comment by yevhen** [comment]
@turowicz Can you share the code?

**Comment by turowicz** [comment]
@ReubenBond can you confirm that there is nothing odd with calling the Grains from the IQueueAdapter?

**Comment by tomachristian** [comment]
@ReubenBond the proposed solution by you and others with the grain queueing a request to itself from inside the callback has a serious flaw, I don’t get why it was proposed… Take the following scenario:

- The timer ticks (the callback delegate begins execution)
- Immediately after this, the grain is scheduled to deactivate (think of it as calling DeactivateOnIdle() as the first thing inside the timer callback to simulate the runtime deactivating the grain)
- The callback awaits a call to the grain itself via a reference
- The call does not get processed by the grain because it was scheduled to deactivate, effectively hanging
- The OnDeactivate method on the grain cannot get called because the timer callback has not completed
- This is a dead-lock

Of course the dead-lock will eventually resolve, as the call to the grain itself from inside the timer callback will eventually timeout, depending on how Orleans is configured.

Also, this call cannot be picked up by another activation as the same grain, because this one (the previous one) has not yet deactivated AFAIK (right?).

A good way to solve this is to not await the call to the grain itself from inside the timer callback, or even better, have that method that gets called be annotated with [OneWay]. Even though this might have the problem of timer ticks getting batched together one after another when the period is really small and the execution of one tick takes longer than the period, due to no longer awaiting.

IMO, this should AT LEAST be properly documented/described and provided with alternatives with pros/cons, even before any fix is added (will it be?).

This has to be done, given that Orleans is supported by MS and is a first class-citizen.

**Comment by Rohansi** [comment]
@tomachristian I've had no issues with the workaround from this thread when using `Task.Run`. Have you tried this https://github.com/dotnet/orleans/issues/2574#issuecomment-475912550?

If you're already doing that then I'm not sure how you'd getting a deadlock at all when timer callbacks don't block calls to the grain. That's the whole issue being raised in this thread - timer callbacks _do not_ block calls to the grain (which isn't intuitive). If they did then that workaround should also trigger a deadlock because the timer callback triggers an external call (separate call chain via `Task.Run`) to the grain and awaits it. How could the grain call run if the timer callback is waiting for the grain call to complete?

**Comment by tomachristian** [comment]
@Rohansi please try this code to see what I mean. I didn't say I have this problem, I am using OneWay to avoid this, but people using the solution proposed here MIGHT have this problem. It is non-deterministic.

```
public interface IMyGrain : IGrainWithStringKey
{
    Task ExecuteSomething();

    Task Tick();
}

public sealed class MyGrain : Grain, IMyGrain
{
    public Task ExecuteSomething()
    {
        var self = this.AsReference<IMyGrain>();

        RegisterTimer(async _ =>
        {
            // Imagine the runtime doing this exactly at this point in time
            DeactivateOnIdle();

            await self.Tick();
        }, null!, TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(1));

        return Task.CompletedTask;
    }

    public Task Tick()
    {
        Console.WriteLine("STUCK BECAUSE OF DEAD-LOCK, EVENTUALLY REACHED BECAUSE OF TIMEOUTS");

        return Task.CompletedTask;
    }

    public override Task OnDeactivateAsync(DeactivationReason reason, CancellationToken cancellationToken)
    {
        Console.WriteLine("STUCK BECAUSE OF DEAD-LOCK, EVENTUALLY REACHED BECAUSE OF TIMEOUTS");

        return Task.CompletedTask;
    }
}

public class Tests : BackgroundService
{
    private readonly IMyGrain _grain;

    public Tests(IGrainFactory grainFactory)
    {
        _grain = grainFactory.GetGrain<IMyGrain>("1");
    }

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        await _grain.ExecuteSomething();
    }
}
```

**Comment by Rohansi** [comment]
@tomachristian I haven't tried that specifically because it is different from the suggested workaround - there is no `Task.Run` around the call to the `Tick` method. While it looks like it would be useless it actually changes how the method is called. Instead of the Orleans runtime seeing the call stack as `ExecuteSomething` -> `Tick` (reentrant) it has separate call stacks for `ExecuteSomething` and `Tick`. The call to `Tick` should have the same semantics as virtual actors in general, if the target grain was deactivated then re-activate it. If that isn't happening then I'd think that's its own bug.

**Comment by tomachristian** [comment]
@Rohansi it dead-locks with and without Task.Run, I invite you to try it.

**Comment by tomachristian** [comment]
It is already reported here https://github.com/dotnet/orleans/issues/7865, btw.

**Comment by Rohansi** [comment]
Good to know. I actually stopped using Orleans because it was so prone to tricky deadlocks. Loved using it but these are some pretty critical issues that have been unresolved for too long.

**Comment by tomachristian** [comment]
@Rohansi indeed. Unfortuantely I am noticing this myself as well, more and more

**Comment by ReubenBond** [comment]
@tomachristian @Rohansi thank you for reporting. I've opened some PRs to fix the deadlock and improve timers in Orleans. The PRs do not yet implement non-reentrant timers, but that is something we hope to address soon. cc @benjaminpetit

**Comment by ReubenBond** [comment]
WIP PR for Non-reentrant timers: #8955

**Comment by turowicz** [comment]
> @turowicz, could you link me to a repro? It may be that something is causing the timer callback passed to `RegisterTimer` to leave the Grain's scheduler, such as `ConfigureAwait(false)`
Dear @ReubenBond, apologies for coming back so late. Are you saying I should add `ConfigureAwait(false)` or remove it? I currently have 0 of this in my code.

**Comment by ReubenBond** [comment]
@turowicz you should not have ConfigureAwait(false) in that code. If you still hit this with v3.7.2 or v8.2.0 (once it's released), please open a new issue and we can investigate together

**Comment by turowicz** [comment]
> @turowicz you should not have ConfigureAwait(false) in that code. If you still hit this with v3.7.2 or v8.2.0 (once it's released), please open a new issue and we can investigate together

We have none of those calls. We are on `7.2.1`.
